json-endpoints/jetpack/class.jetpack-json-api-sync-endpoint.php

```diff
@@ -233,3 +239,151 @@ protected function result() {
+			$codec = new Jetpack_Sync_JSON_Deflate_Array_Codec();
+			$codec_name = $codec->name();
+		}
+
+		// We need the sender to set up all the before send module actions...
+		require_once JETPACK__PLUGIN_DIR . 'sync/class.jetpack-sync-sender.php';
+		$sender = Jetpack_Sync_Sender::get_instance();
+		
+		$skipped_items_ids = array();
+
+		// let's delete the checkin state
+		if ( isset( $args['force'] ) && $args['force'] ) {
+			$queue->force_checkin();
+		}
+
+		$buffer = $this->get_buffer( $queue, $args[ 'number_of_items' ] );
+		
+		// Check that the $buffer is not checkout out already
+		if ( is_wp_error( $buffer ) ) {
+			return new WP_Error( 'buffer_open', "We couldn't get the buffer it is currently checked out", 400 );
+		}
+		
+		if ( ! is_object( $buffer ) ) {
+			return new WP_Error( 'buffer_non-object', 'Buffer is not an object', 400 );
+		}
+
+		$items = $buffer->get_items();
+
+		// set up current screen to avoid errors rendering content
+		require_once( ABSPATH . 'wp-admin/includes/class-wp-screen.php' );
+		require_once( ABSPATH . 'wp-admin/includes/screen.php' );
+		set_current_screen( 'sync' );
+
+		foreach ( $items as $key => $item ) {
+			// Suspending cache addition help prevent overloading in memory cache of large sites.
+			wp_suspend_cache_addition( true );
+
+			/** This filter is documented in sync/class.jetpack-sync-sender.php */
+			$item[1] = apply_filters( 'jetpack_sync_before_send_' . $item[0], $item[1], $item[2] );
+			wp_suspend_cache_addition( false );
+			if ( $item[1] === false ) {
+				$skipped_items_ids[] = $key;
+				continue;
+			}
+			$items_to_send[ $key ] = ( $encode ?  $codec->encode( $item ) : $item );
+		}
+	
+		return array(
+			'buffer_id'      => $buffer->id,
+			'items'          => $items_to_send,
+			'skipped_items'  => $skipped_items_ids,
+			'codec'          => $codec_name,
+			'sent_timestamp' => time(),
+		);
+	}
+
+	protected function get_buffer( $queue, $number_of_items ) {
+		$start = time();
+		$max_duration = 5;
+
+		$buffer = $this->try_getting_buffer( $queue , $number_of_items );
+		$duration = time() - $start;
+
+		while( ! $buffer && $duration < $max_duration ) {
+			$buffer = $this->try_getting_buffer( $queue , $number_of_items );
+			$duration = time() - $start;
+		}
+
+		return $buffer;
+	}
+
+	protected function try_getting_buffer( $queue , $number_of_items ) {
+		$buffer = $queue->checkout( $number_of_items );
+		if ( is_wp_error( $buffer ) ) {
+			sleep( 2 ); // let's wait for 2 seconds before we even allow this function to be called again.
+		}
+		return $buffer;
+	}
+}
+
+class Jetpack_JSON_API_Sync_Close_Endpoint extends Jetpack_JSON_API_Sync_Endpoint {
+	protected function result() {
+		$request_body = $this->input();
+		$queue_name = $this->validate_queue( $request_body['queue'] );
+
+		if ( is_wp_error( $queue_name ) ) {
+			return $queue_name;
+		}
+		require_once JETPACK__PLUGIN_DIR . 'sync/class.jetpack-sync-queue.php';
+
+		if ( ! isset( $request_body['buffer_id'] ) ) {
```

Comment:
can't we just make this a required parameter in the API definition? that way it'll return a validation error without ever calling result()